### ISSUE TITLE
[3.x] Requirements::includeInHTML inserts things into more than one </head> tag

### ISSUE THREAD

**Issue by robbieaverill** [silverstripe/silverstripe-framework#6779]
For example: if I insert a third-party Javascript snippet (like an A/B tester, etc) into my page via:

```php
Requirements::customScript('//.....');
```

... and my custom script has a _string literal_ like `var whatever = "<head></head>"`, then `Requirements::includeInHTML` will inject requirements into that `</head>` as well as the closing tag that is actually part of the DOM. This theoretically would also affect code documentation examples that have a `</head>` in them (that aren't HTML escaped).

Adding a limit of 1 to the `preg_replace` calls seems safe and fixes the problem (unless you had the same thing happening in the head tag itself).

**Comment by tractorcow** [comment]
I suggest moving away from regexp for body manipulation and use DOMDocument. http://stackoverflow.com/questions/12378675/adding-style-tags-to-head-with-php-domdocument

**Comment by robbieaverill** [comment]
Concerns:

* Performance
* DOMDocument doing unwanted things to the page

**Comment by sminnee** [comment]
I suggest moving away from inline scripts ;-)

**Comment by sminnee** [comment]
We couldn't afford more than a few ms for the operation so I agree with perf concerns.

As well as avoiding custom scripts generally, I'd suggest refactoring to replace only the first instance of </head>

**Comment by robbieaverill** [comment]
>I'd suggest refactoring to replace only the first instance of /head

So that's what #6780 does, but if you don't "force JS to bottom", which you don't by default in SS, then a snippet in the head tag which contains `</head>` will still be the first occurrence and it'll actually cause more problems than it currently does.

In my case I'm forcing JS to bottom so the snippet isn't the first occurrence and limiting the replacements works for me, but as @tractorcow suggested I've just modified my JS snippet so it has `</he" + "ad>` instead and doesn't get picked up by the regex.

---

This area does look like it could be improved to be honest, but without increasing load time slightly using DOM document I'm not sure what the best solution is...

**Comment by sminnee** [comment]
@robbieaverill The underlying issue is "why are we doing more than one search & replace pass?" There should be exactly 1 execution of code that looks for a `</head>` and writes into it, and everything else that wants to use this should feed into it.

**Comment by sminnee** [comment]
Is it as simple as just refactoring these 2 lines into one?

Requirements.php:904
```php
				$content = preg_replace("/(<\/head>)/i", $requirements . "\\1", $content);
				$content = preg_replace("/(<\/head>)/i", $jsRequirements . "\\1", $content);
```

**Comment by robbieaverill** [comment]
PR reopened at #6780

**Comment by robbieaverill** [comment]
Resolved in #6780 for 3.5+
